src/main/java/gross/gameoflife/Grid.java

```diff
@@ -0,0 +1,224 @@
+        ArrayList<int[]> cellsToComeAlive = new ArrayList<>();
+
+        // traverse the entire 2D array to identify all living cells
+        for (int i = 0; i < board.length; i++) {
+            for (int j = 0; j < board[i].length; j++) {
+                int cell = board[i][j];
+                // if cell is alive, check status of neighbors
+                if (cell == 1) {
+                    // identify all live neighbors
+                    int liveNeighbors = findLiveNeighbors(i, j);
+                    // will die if it does not have 2 or 3 neighbors (1 or less, 4 or more)
+                    if (liveNeighbors != 2 && liveNeighbors != 3) {
+                        int[] cellLocation = {i, j};
+                        cellsToDie.add(cellLocation);
+                    }
+                    // identify all dead neighbors
+                    ArrayList<int[]> deadNeighbors = findDeadNeighbors(i, j);
+                    // loop through all dead cells, check if they have 3+ neighbors
+                    for (int[] deadCell : deadNeighbors) {
+                        boolean alreadyUsed = false;
+                        for (int[] c : cellsToComeAlive) {
+                            if (c[0] == deadCell[0] && c[1] == deadCell[1]) {
+                                alreadyUsed = true;
+                            }
+                        }
+                        if (!alreadyUsed) {
+                            // add to cellsToComeAlive if not already there
+                            if (findLiveNeighbors(deadCell[0], deadCell[1]) == 3) {
+                                cellsToComeAlive.add(deadCell);
+                            }
+                            // can be optimized by tracking which cells were already checked
+                        }
+                    }
+                }
+            }
+        }
+        // kill all cells in cellsToDie
+        for (int[] cellToKill : cellsToDie) {
+            setCellDead(cellToKill[0], cellToKill[1]);
+        }
+
+        // revive all cells in cellsToComeAlive
+        for (int[] cellToRevive : cellsToComeAlive) {
+            setCellAlive(cellToRevive[0], cellToRevive[1]);
+        }
+    }
+
+    public int findLiveNeighbors(int row, int column) {
+        int liveNeighbors = 0;
+
+        boolean hasTop = row > 0; // a higher row exists
+        boolean hasBottom = row < board.length - 1; // a lower row exists
+        boolean hasLeft = column > 0; // a left column exists
+        boolean hasRight = column < board[0].length - 1; // a right column exists
+
+        if (hasTop) {
+            // check if cell directly above current cell is alive
+            if (board[row - 1][column] == 1) {
+                liveNeighbors++;
+            }
+            if (hasLeft) {
+                // check if diagonal upper left cell is alive
+                if (board[row - 1][column - 1] == 1) {
+                    liveNeighbors++;
+                }
+            }
```

Comment:
Combine these two if statements.
Do the same for the other if statements in this method.